flipperdevices/flipperzero-firmware
# ble: profile rework

# What's new

**General BLE**:
- ble: reworked profile subsystem
- now applications can define custom BLE profiles
- profiles define BLE advertising parameters and own service set

**BLE internals**:
- combined SHCI & HCI event processing threads
- reworked event dispatching to services to support new way of defining profiles and overcome fixed wpan handler count limit
- battery service (&others): now supports multiple instances
- bt service: now only init BLE pairing dialog when needed to save memory
- moved BLE HID profile from furi_hal to lib/ble_profile as an example (custom profiles can also be defined in app's private libs)

**Extra beacon**:
- added API for controlling additional, non-connectable BLE advertisements (BLE beacon)
- that beacon works in background, independently from main BLE subsystem
- extra beacon has status indication on desktop
- added example app for configuring and testing extra beacon feature

**Misc**:
- Moved example_custom_font app from debug to examples
- Improved "ps" CLI command output: added thread priorities
- Added FURI_PACKED macro for packed structures
- hid_app: cleaner code split for USB/BLE transports
- fbt: better handling of malformed target.json files

# Verification 

- test that update from older versions and downgrade do not break BLE pairing
- run BLE remote app, check that old pairings work & app itself works
- run BLE stress test from companion apps
- run BLE Beacon app in various combinations with BLE on/off, RPC active/not
- check out cool "ps"
- idk

# Checklist (For Reviewer)
